juju-tutorials/juju-tricks.md

```diff
@@ -0,0 +1,177 @@
+{'cni.configured': None,
+ 'cni.connected': None,
+ 'cni.is-master': None,
+ 'endpoint.cni.changed.cni-conf-file': None,
+ 'endpoint.cni.changed.egress-subnets': None,
+ 'endpoint.cni.changed.ingress-address': None,
+ 'endpoint.cni.changed.is_master': None,
+ 'endpoint.cni.changed.private-address': None,
+ 'endpoint.cni.joined': None,
+ 'etcd.available': {'conversations': ['reactive.conversations.etcd.global'],
+                    'relation': 'etcd'},
+ 'etcd.connected': {'conversations': ['reactive.conversations.etcd.global'],
+                    'relation': 'etcd'},
+ 'etcd.tls.available': {'conversations': ['reactive.conversations.etcd.global'],
+                        'relation': 'etcd'},
+ 'flannel.binaries.installed': None,
+ 'flannel.cni.available': None,
+ 'flannel.etcd.credentials.installed': None,
+ 'flannel.network.configured': None,
+ 'flannel.service.installed': None,
+ 'flannel.service.started': None,
+ 'flannel.version.set': None}
+```
+
+## Accessing juju internal database
+
+Duration: N/A
+
+[note status="Caution"]
+Changing values in the juju database can corrupt your juju installation. Proceed with caution and doublecheck before making any changes.
+[/note]
+
+Sometimes it may be necessary to access juju's database directly, although it should be considered as last resort solution. To log into the juju's database, you must first log into the juju controller machine.
+
+```console
+juju ssh -m controller 0
+```
+
+Once there, you can log into the juju's mongo database.
+
+```console
+mongo --sslAllowInvalidCertificates --authenticationDatabase admin --ssl -u $(sudo awk '/tag/ {print $2}' /var/lib/juju/agents/machine-?/agent.conf) -p $(sudo awk '/statepassword/ {print $2}' /var/lib/juju/agents/machine-?/agent.conf) localhost:37017/juju
+```
+
+### Manual change of Openstack endpoint certificates
+
+In case you are running juju controller on top of Openstack and for some reason the CA certificates on the endpoints change, you can't access the endpoint anymore with "certificate signed by unknown authority", the only way to fix it is to update the juju mongodb directly.
+
+Start by logging into the controller node and then into the mongo interactive shell. Once you are in the mongo shell, update certificate field for your cloud in `clouds` collection.
+
+```console
+juju:PRIMARY> db.clouds.update({"_id": "openstack_cloud"}, {$set: {"ca-certificates": [""]}})
+```
+
+Certificate must be PEM-formatted one-liner with newlines replaced by `\n`.
+
+`-----BEGIN CERTIFICATE-----\n<CERTIFICATE_CONTENT>\n\n-----END CERTIFICATE-----\n`
+
+### Debugging stuck transactions
+
+If a transaction is stuck and you want to figure out what's going on, look at the 'cleanups' collection.
+
+```console
+juju:PRIMARY> db.cleanups.find().pretty()
+```
+
+### Dumping database content
+
+There already is `juju create-backup` tool, but in case it's not working or you just want to dump raw juju  database, you can log into the controller and run following.
```

Comment:
"Although `juju create-backup` exists, you may want to dump the Juju database in raw by logging into the controller and running the following:"

Reply:
Maybe even

"Although `juju create-backup` exists, you can dump raw Juju database by logging into the controller and running the following:"

I feel like the part "you may want to" makes it sound like it's something recommended, to me it sounds like "It may be a good idea to dump raw juju database".